File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/RetryWithDelay.java
package com.github.kongpf8848.rxhttp;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.functions.Function;

public class RetryWithDelay implements Function<Observable<Throwable>, ObservableSource<?>> {

    private final int maxRetries;
    private final long retryDelayMillis;
    private int retryCount;

    public RetryWithDelay(final int maxRetries, final long retryDelayMillis) {
        this.maxRetries = maxRetries;
        this.retryDelayMillis = retryDelayMillis;
        this.retryCount = 0;
    }

    @Override
    public ObservableSource<?> apply(Observable<Throwable> throwableObservable) throws Exception {
        return throwableObservable.flatMap(new Function<Throwable, ObservableSource<?>>() {
            @Override
            public ObservableSource<?> apply(Throwable throwable) throws Exception {
                if (++retryCount <= maxRetries) {
                    return Observable.timer(retryDelayMillis, TimeUnit.MILLISECONDS);
                }
                return Observable.error(throwable);
            }
        });
    }
}

File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/RetryWithDelay.kt
package com.github.kongpf8848.rxhttp

import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.functions.Function
import java.util.concurrent.TimeUnit

class RetryWithDelay(private val maxRetries: Int, private val retryDelayMillis: Long) : Function<Observable<Throwable?>, ObservableSource<*>> {
    private var retryCount = 0

    @Throws(Exception::class)
    override fun apply(throwableObservable: Observable<Throwable?>): ObservableSource<*> {
        return throwableObservable.flatMap(object : Function<Throwable?, ObservableSource<*>> {
            @Throws(Exception::class)
            override fun apply(throwable: Throwable): ObservableSource<*> {
                return if (++retryCount <= maxRetries) {
                    Observable.timer(retryDelayMillis, TimeUnit.MILLISECONDS)
                } else Observable.error<Any>(throwable)
            }

        })
    }

}